Use cscore for PublishVideoOperation

SamCarlberg opened the pull request:
Updates JavaCV to 1.3.3 and javacpp-opencv to 3.2.0 in order to keep the same memory model between bindings.

Allows up to ten publish video operations on ports 1180 to 1189.

Note: this increases the size of the core jar by ~67MB from bundling the cscore and opencv binaries.

#### GRIP pipeline
![grip-cscore-pipeline](https://user-images.githubusercontent.com/6320992/32620093-b80ced2c-c54a-11e7-9061-141abaaa771b.PNG)

#### Video streams in Shuffleboard using cscore
![grip-streams-in-shuffleboard](https://user-images.githubusercontent.com/6320992/32620094-b81aa43a-c54a-11e7-97e4-4e9106a88d16.PNG)

SamCarlberg committed:
Use cscore for PublishVideoOperation

Updates JavaCV to 1.3.3 and javacpp-opencv to 3.2.0

SamCarlberg committed:
Smart port reuse

SamCarlberg commented on build.gradle:
```diff
@@ -225,10 +225,10 @@ project(":core") {
+        compile group: 'org.bytedeco', name: 'javacv', version: '1.3.3'
+        compile group: 'org.bytedeco.javacpp-presets', name: 'opencv', version: '3.2.0-1.3'
+        compile group: 'org.bytedeco.javacpp-presets', name: 'opencv', version: '3.2.0-1.3', classifier: os
+        //compile group: 'org.bytedeco.javacpp-presets', name: 'opencv-3.0.0-1.1', classifier: 'linux-frc'
```
Commented out because there's no `opencv-3.2.0-1.3` with the `linux-frc` classifier and I'm worried about different memory models between binding versions (JavaCV with 3.0.0, OpenCV with 3.2.0) . This could get removed entirely if we remove the deploy functionality.

JLLeitschuh replied:
With this commented out as it is is the deploy functionality essentially nerfed? If so, then we should remove the deployer for this season.

SamCarlberg committed:
Improve NetworkTable code

JLLeitschuh commented:
Travis fails, this is a non-starter for me.
We need to introduce the docker based build system to resolve this first before this gets merged.

JLLeitschuh commented on core/src/main/java/edu/wpi/grip/core/operations/composite/PublishVideoOperation.java:
```diff
@@ -167,25 +125,51 @@ public PublishVideoOperation(InputSocket.Factory inputSocketFactory) {
+      try {
+        Field nativeObjField = Mat.class.getField("nativeObj");
+        nativeObjField.setAccessible(true);
+        nativeObjField.setLong(openCvMat, javaCvMat.address());
```
This seems like a terrible idea, but I understand what you are doing here.
Do you actually see speed benefits from doing this?
@saudet Is there any reason why doing this assignment could shoot us in the foot?

This is taking the memory address of the OpenCV mat and basically telling JavaCV to use it's memory instead.

saudet replied:
First it'd say, what is missing from JavaCV that you need to use the other set of bindings? IMO, we should fix that so we don't need to do things like that.

And even when that's necessary, we don't need to use reflection.

Just passing the pointer address around is sufficient, as long as you make sure they are from the same version of OpenCV, compiled with the same compiler, the same flags, etc and to make sure memory isn't prematurely deallocated and that it gets deallocated by the same binaries used to allocate it.

JLLeitschuh replied:
Is there a way of creating a `Mat` just using the memory address??

JLLeitschuh replied:
@SamCarlberg Ping on this? Thoughts on making sure the memory isn't inconsistently GCed?

ThadHouse replied:
See my comment at the bottom. In the rewrite, all the memory mapping between different opencv versions will be gone. There will be no wpilib opencv interface anymore.

ThadHouse replied:
As an example, The following class would be added to GRIP. It will use the JavaCV version of OpenCV, and always copy from the raw image into the requested Mat. It is memory safe, and does not depend at all on the internal representation of how an OpenCV Mat looks. There's some tricks on the cscore jni side to avoid allocations when needed, but everything is in spec, no reflection or undefined behavior.

https://github.com/wpilibsuite/allwpilib/blob/13e15663b6570028a9222b4990d141d16fda3a72/cscore/src/dev/java/edu/wpi/cscore/RawCVMatSink.java

JLLeitschuh replied:
@SamCarlberg Do we want to wait for that update then?

SamCarlberg replied:
I think so. Thad's offered to do do the rewrite with the new cscore stuff. Maybe we should also use the cameraserver library for doing the stream hosting instead of manually doing the Networktable side

ThadHouse replied:
I wouldn't be able to bring it in directly (because of the opencv dependency) but I could bring in most of it. There's a lot extra we wouldnt need there as well.

ThadHouse replied:
I wouldn't be able to bring it in directly (because of the opencv dependency) but I could bring in most of it. There's a lot extra we wouldnt need there as well.

JLLeitschuh commented on core/src/main/java/edu/wpi/grip/core/operations/composite/PublishVideoOperation.java:
```diff
@@ -46,110 +59,55 @@
+
+  // Write to the /CameraPublisher table so the MJPEG streams are discoverable by other
+  // applications connected to the same NetworkTable server (eg Shuffleboard)
+  private static final ITable cameraPublisherTable = NetworkTable.getTable("/CameraPublisher");
```
Can you not load this statically? Instead do it at instantiation time.

SamCarlberg replied:
What would that gain?

JLLeitschuh replied:
If it fails then GRIP doesn't crash immediately.

SamCarlberg replied:
So instead of crashing to desktop at initialization when creating the operations list, it'll crash to desktop when adding a new PublishVideo operation. I guess that's better because at least users will see it happens when adding the operation

JLLeitschuh commented on core/src/main/java/edu/wpi/grip/core/operations/composite/PublishVideoOperation.java:
```diff
@@ -10,34 +10,47 @@
+      logger.log(Level.SEVERE, "CameraServerJNI load failed! Exiting", e);
+      System.exit(31);
+    }
+  }
```
Does this need to happen statically? Can't this be done when the operation is constructed?

SamCarlberg commented on core/src/main/java/edu/wpi/grip/core/operations/composite/PublishVideoOperation.java:
```diff
@@ -10,34 +10,47 @@
+      CameraServerJNI.getHostname();
+    } catch (Throwable e) {
+      logger.log(Level.SEVERE, "CameraServerJNI load failed! Exiting", e);
+      System.exit(31);
```
This should set a flag and make the operation throw an exception in `perform()` instead of suddenly exiting the entire application

JLLeitschuh replied:
Yea, we want to limit the number crashes that don't display the crash dialog.

ThadHouse replied:
Theres actually nothing to catch, because internally in the JNI loading classes we call System.exit(1). There would be no exception ever caught here. It would be impossible to catch.

ThadHouse replied:
After a little discussion, we will leave the IO Exception that the static constructor can throw in flight, that way its catchable. That will be a change on our side. We will do the same thing to ntcore as well.

SamCarlberg committed:
Updates from review. Add lots of documentation, fix hostname on windows

Set a flag when cscore can't be loaded to make the operation perform() fail, instead of crashing to desktop
Still need to do mac hostname resolution

JLLeitschuh commented on core/src/main/java/edu/wpi/grip/core/operations/composite/PublishVideoOperation.java:
```diff
@@ -167,25 +142,109 @@ public PublishVideoOperation(InputSocket.Factory inputSocketFactory) {
+   * {@link CameraServerJNI#getHostname() getHostName() function} only works on Linux, so we need to
+   * implement the method for Windows and Mac ourselves.
+   */
+  private static String getHostName() {
```
Replace with: `InetAddress.getLocalHost().getHostName()`???

SamCarlberg replied:
Oooh. That's much better 👍

JLLeitschuh commented on core/src/main/java/edu/wpi/grip/core/operations/composite/PublishVideoOperation.java:
```diff
@@ -167,25 +142,109 @@ public PublishVideoOperation(InputSocket.Factory inputSocketFactory) {
+
+    if (!cscoreLoaded) {
+      throw new IllegalStateException(
+          "cscore could not be loaded. The image streaming server cannot be started.");
```
👍

JLLeitschuh commented on core/src/main/java/edu/wpi/grip/core/operations/composite/PublishVideoOperation.java:
```diff
@@ -10,146 +10,121 @@
-  private static final int PORT = 1180;
-  private static final byte[] MAGIC_NUMBER = {0x01, 0x00, 0x00, 0x00};
+  private static final int INITIAL_PORT = 1180;
+  private static final int MAX_STEP_COUNT = 10; // limit ports to 1180-1189
```
👍

JLLeitschuh commented on core/src/main/java/edu/wpi/grip/core/operations/composite/PublishVideoOperation.java:
```diff
@@ -167,25 +142,109 @@ public PublishVideoOperation(InputSocket.Factory inputSocketFactory) {
+    serverSource.setResolution(input.size().width(), input.size().height());
+    serverSource.putFrame(publishMat);
+    if (lastFrame != -1) {
+      long dt = now - lastFrame;
```
I've seen this actually be `0` even though you are using nano seconds. (This is highly unlikely but we saw it in the camera source code in one bug) This will cause a DIV by zero exception.

SamCarlberg replied:
It'll actually be `Double.Infinity` per IEEE spec for floating-point numbers (`1e9 / 0L = Infinity`)

JLLeitschuh replied:
`(int) 1e9 / 0L` will throw `java.lang.ArithmeticException: / by zero` because of the cast.

JLLeitschuh replied:
In kotlin: `(1e9 / 0L).toInt()` is equal to `2147483647` which is an insanely high frame rate. Is that what you want to set it to?

SamCarlberg replied:
It's technically infinite since the new frame comes at the same time as the previous one. But it doesn't happen here because it's rate limited by the pipeline and the update rate of the pipeline sources. I seriously doubt `dt` will ever be less than a few hundred microseconds.

JLLeitschuh replied:
Okay, that's fine then.

JLLeitschuh commented on core/src/main/java/edu/wpi/grip/core/operations/composite/PublishVideoOperation.java:
```diff
@@ -10,146 +10,121 @@
+      // Loading the CameraServerJNI class will load the appropriate platform-specific OpenCV JNI
+      CameraServerJNI.getHostname();
+      loaded = true;
+    } catch (Throwable e) {
```
Can native bindings ever throw an `InterruptedException`? If so, make sure that this handles it correctly.

SamCarlberg replied:
It's most likely to be `UnsatisfiedLinkError` or a related error. No interrupted exceptions here.

ThadHouse replied:
Unless we explicitly throw interrupted from the JNI level, nothing in JNI otherwise would throw it. And since its a dumb idea to interrupt a thread anyway, we would never throw it internally. Pretty much all that shows up is UnsatisfiedLinkError, or one of the few runtime exceptions we throw.

Note this should be using forceLoad(), not getHostname() (I don't think we had forceLoad() then), but I will my sure my version does that.

JLLeitschuh commented on core/src/main/java/edu/wpi/grip/core/operations/composite/PublishVideoOperation.java:
```diff
@@ -167,25 +142,109 @@ public PublishVideoOperation(InputSocket.Factory inputSocketFactory) {
+   * @param openCvMat the OpenCV Mat wrapper object to copy into
+   * @throws RuntimeException if the OpenCV native pointer could not be set
+   */
+  private static void copyJavaCvToOpenCvMat(opencv_core.Mat javaCvMat, Mat openCvMat)
```
Can you add some unit tests for this function to make sure that it gets covered. I don't want someone in the future to be updating dependencies and have this break because this code doesn't get run by a test.

SamCarlberg committed:
Use InetAddress.localhost() for getting hostname, add test for mat copy

SamCarlberg commented on core/src/main/java/edu/wpi/grip/core/operations/composite/PublishVideoOperation.java:
```diff
@@ -7,149 +7,130 @@
+        InetAddress localHost = InetAddress.getLocalHost();
+        ourTable.putStringArray("streams",
+            new String[]{
+                generateStreamUrl(localHost.getHostName(), ourPort),
```
@PeterJohnson Would it be prudent to add a ".local" to the end of the local hostname e.g. `team-190-driverstation.local`? I'm not sure how mDNS would handle it on an FRC field

PeterJohnson replied:
Note `InetAddress.getLocalHost()` may not give you the address you want if the machine has multiple network adapters (https://stackoverflow.com/questions/5813194/inetaddress-getlocalhost-does-not-return-expected-ip-address-from-c-windows-s).  You probably want to use `NetworkInterface.getNetworkInterfaces()` instead.  cscore provides the similar `CameraServerJNI.getNetworkInterfaces()` as well as the NetworkInterfacesChanged event for listening to network interface changes, but at the moment these only work on Linux.  In either case, appending ".local" is almost certainly the wrong thing to do; you should publish the raw IP instead; appending ".local" will require the computer running GRIP to publish its mDNS address, etc.

The streams entry is an array for exactly this reason, so you can e.g. generate an entry for each network address the machine has (e.g. on the robot side we publish with both the USB and Ethernet IP addresses).

ThadHouse replied:
getNetworkInterfaces works on windows and mac now, so we will be able to use the cscore functions for that. My rewrite will do this

JLLeitschuh commented on core/src/test/java/edu/wpi/grip/core/operations/composite/PublishVideoOperationTest.java:
```diff
@@ -0,0 +1,74 @@
+  @BeforeClass
+  public static void initialize() {
+    // Make sure the OpenCV JNI is loaded
+    blackHole(PublishVideoOperation.DESCRIPTION);
```
Derpy but sure.

JLLeitschuh commented on core/src/test/java/edu/wpi/grip/core/operations/composite/PublishVideoOperationTest.java:
```diff
@@ -0,0 +1,74 @@
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertThat;
+
+public class PublishVideoOperationTest {
```
👍 Thanks for adding this test for us!

JLLeitschuh commented on core/src/main/java/edu/wpi/grip/core/operations/composite/PublishVideoOperation.java:
```diff
@@ -115,8 +113,16 @@ public PublishVideoOperation(InputSocket.Factory inputSocketFactory) {
+                generateStreamUrl(localHost.getHostAddress(), ourPort)
+            });
+      } catch (UnknownHostException e) {
+        ourTable.putStringArray("streams", new String[0]);
```
Should we log something here?

SamCarlberg replied:
Maybe? The documentation isn't very clear on what circumstances will cause this exception to be thrown. It might never be thrown for the local host since it appears to only happen if the host is unreachable.

JLLeitschuh replied:
Interesting. @PeterJohnson do you have any thoughts on what might cause this on the local system or if this is a non-issue?

PeterJohnson replied:
It sounds like how the Java getLocalHost function works is it gets the local hostname, then resolves it into an address.  So it's definitely possible to fail: https://stackoverflow.com/questions/1881546/inetaddress-getlocalhost-throws-unknownhostexception

saudet commented:
https://docs.opencv.org/java/2.4.2/org/opencv/core/Mat.html#Mat(long) ?

JLLeitschuh commented:
@SamCarlberg The above solves our problem.

SamCarlberg commented:
Well, kind of. I'm still wary about binary incompatibilities between the JavaCV builds and the WPI OpenCV builds. The perfect solution would be to have cscore accept raw image data (bytes, width, height) to avoid needing to use the OpenCV bindings altogether. Something like manually creating [cscore images](https://github.com/wpilibsuite/cscore/blob/master/src/main/native/cpp/Image.h).

That'd let us keep the OpenCV bindings as a runtime dependency for cscore instead of a compile-time dependency for GRIP.

PeterJohnson commented:
I'm working on a raw image API for cscore for exactly that reason.

saudet commented:
We can still share the data of the image without sharing the whole Mat
struct. That would be a much better solution.

SamCarlberg commented:
Doesn't that still run the risk of binary incompatibilities?

saudet commented:
Just sharing image data? No worries at all, just need to make sure
deallocation is done right.

saudet commented:
On Windows though, symbols are loaded globally for the process, but I think
the Java bindings from OpenCV don't expose the C++ API anyway, so should be
alright.

SamCarlberg committed:
Use Mat ctor for copying pointer, use NetworkInterface to get host IPs

JLLeitschuh commented on core/src/main/java/edu/wpi/grip/core/operations/composite/PublishVideoOperation.java:
```diff
@@ -160,7 +171,16 @@ public void perform() {
+    // object that gets copied into.  The data address will change, however, if the image is resized
+    // or changes type.
+    if (publishMat == null || publishMat.nativeObj != input.address()) {
+      publishMat = new Mat(input.address());
```
Can you keep the test you had for `copyJavaCvToOpenCvMat ` arround for making sure that we know if the memory models are incompatible when we bump versions in the future?

SamCarlberg replied:
Sure.

SamCarlberg committed:
Re-add test for compatiblity between JavaCV and OpenCV

Fix an issue with codegen tests from the JavaCV update

SamCarlberg committed:
Remove unused static list of network interfaces, use assume() in test

SamCarlberg committed:
Copy bytes instead of native pointers

Should fix problems resulting from incompatible binaries

SamCarlberg committed:
Merge branch 'master' into cscore-publishing

SamCarlberg committed:
Remove unneeded preloading

JLLeitschuh commented:
Wow, serious throwback to an old PR. Thanks @saudet!

ThadHouse commented:
I'm probably going to completely rewrite most of this using the RawFrame API I'm adding to CScore, but I want to leave it open for now for reference for some of the more weird ui stuff that I don't know.

ThadHouse commented:
@SamCarlberg do you want to move the unrelated stuff out of this PR into another PR? Like the DoublePointer stuff? Or has that already been done in other PR's?

JLLeitschuh commented:
@SamCarlberg is this PR outdated now that the one from @ThadHouse is merged?

ThadHouse commented:
My PR only did inputs, not output.
